### ISSUE TITLE
[Feature request] - Check expired token for refreshing it

### ISSUE THREAD

**Issue by cs-victor-nascimento** [bryanjos/joken#31]
It is common the need for refreshing a JWT token. Once an issued token expires the client should be able to renew it with some mechanism. Refreshing is not Joken's purpose but in this scenario we need to check the validity of a token even if it has expired. 

The issue is that decoding a token with Joken will give an error in the first claim (exp) that will be piped through all other claims without checking. So, it would be nice to have a way to decode a token and know that all claims BUT exp are valid and refreshing can proceed.

This is perhaps a new function in Joken module that would delegate to Joken.Token.decode with some kind of option. Something like:

```elixir
# this enforces the caller to know the token may be expired
{:expired, payload} = Joken.decode_expired some_jwt
```

Or maybe some optional parameter to Joken.decode like:

```elixir
# optional skip parameter at call site that shows the caller knows what he is doing
{:ok, payload} = Joken.decode some_jwt, skip: [:exp]
```
Personally, I prefer the second option but that is debatable. If desired I can come up with a PR for these options.

**Comment by bryanjos** [comment]
I like the second option as well. 

I have been thinking about issue #26 a lot which may help for what you want as well. The idea would be to allow the adding and validation of claims to be overridable.

Although maybe the second option you proposed would still be a good idea for one-off situations. I'm open to ideas on any of it.

**Comment by cs-victor-nascimento** [comment]
I see #26 more as a way to define custom validation for claims. Here is just a way to bypass validation at all for a given set of claims. I believe both are nice features and are related :+1: 

With the second option we coud add it without impacting on current users of the library. First I would add an optional parameter to Joken.decode that would be a keyword list. This would then be passed to Joken.Token.decode and it would be passed to all Claim checks. That is where #26  would kick in! The decode behaviour callback could accept the options list.

Do you think we should have a PR with both features backed together? Maybe it is to wait for the behaviour implementation before going in with this one. I will ad a comment on #26  too.

**Comment by bryanjos** [comment]
I think this can be done separately from #26. Just playing around a bit in the code, one approach would be to have a keyword list with the key being the claim and the value being the function used for validation. Something like this:

```elixir
    validations = [
      exp: &Claims.check_exp/1, 
      nbf: &Claims.check_nbf/1, 
      iat: &Claims.check_iat/1,
      aud: &Claims.check_aud/2,
      iss: &Claims.check_iss/2,
      sub: &Claims.check_sub/2
    ]
```

Then filtering out the items based on the keys in the skip parameter. From there mapping the remaining items, calling the associated function and then filtering to only return the errors

```elixir
        results = Enum.reduce(skip, validations, fn(key, current_validations) ->
          Keyword.delete(current_validations, key)
        end)
        |> Enum.map(fn({key, func}) ->
          func.(data)
        end)
        |> Enum.filter(fn({key, value}) -> 
          key == :error
        end)
        |> Enum.map(fn(x) -> elum(x, 1) end)
```

if there aren't any errors, then `{:ok, data}` can be returned, else return `{:error, list_of_validation_error_messages}`

The only breaking change in this situation being that on error it returns a list of messages instead of just a string message.

**Comment by cs-victor-nascimento** [comment]
I thought about the behaviour before because skip is one option. Perhaps there will be other options in the future and processing those would make this code a bit more convoluted. Though I agree that perhaps is better to wait that happen... 

I think that a list of errors may not be that helpful for clients of the library. To know if a specific claim failed people would have to match on binaries. 

I think we could stop processing when an error was found and maintain current behaviour. A Stream could be handy here.

**Comment by bryanjos** [comment]
I should have some time soon to take a look at both and see where everything fits. You're right it does make more sense to stop at the first error.

**Comment by cs-victor-nascimento** [comment]
Fixed on #36.
